torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,1010 @@
+    norm_type = float(norm_type)
+    params_with_grad = self.params_with_grad
+    # Computes the max norm for this shard's gradients and sync's across workers
+    local_norm = _calc_grad_norm(params_with_grad, norm_type)
+    if norm_type == inf:
+      total_norm = xm.all_reduce(xm.REDUCE_MAX, local_norm)
+    else:
+      total_norm = xm.all_reduce(xm.REDUCE_SUM, local_norm**norm_type)
+      total_norm = total_norm**(1.0 / norm_type)
+
+    # Now multiply each grad by (max_norm/total_norm), same as torch 1.7 https://tinyurl.com/3wtxhhqq)
+    clip_coef = torch.clip(max_norm / (total_norm + 1e-6), 0.0, 1.0)
+    for p in params_with_grad:
+      p.grad.detach().mul_(clip_coef.to(p.grad.device))
+
+    return total_norm
+
+  @torch.no_grad()
+  def _shard_parameters_(self, params_to_shard) -> None:
+    """
+    At initialization we wrap a module with full parameters and shard the
+    parameters in-place. Sharding is implemented by viewing each parameter
+    as a 1D Tensor and retaining only a single slice, where the slice size
+    is determined by the number of data parallel workers.
+
+    Wrapping modules with many small parameters (or with a very large data
+    parallel world size) will result in many small parameter shards and slow
+    performance. In this case it's better to set *``flatten_parameters``* to
+    ``True``, so that all of the small parameters in the module are combined
+    into a single contiguous Tensor and sharded once.
+
+    After this initial sharding is complete, the user can initialize a
+    ``torch.optim.Optimizer`` in the usual way, i.e.::
+
+    .. code-block:: python
+
+        optim = torch.optim.Adam(sharded_module.parameters(), lr=0.0001)
+
+    The optimizer will see only a single slice of parameters and will thus
+    allocate less memory for optimizer state, avoiding redundancy across
+    data parallel workers.
+
+    Note: this method is implemented in a different manner from
+    ``fairscale.nn.FullyShardedDataParallel``. Here we delete the original
+    module parameters and create new sharded parameter tensors (instead of
+    making sharded tensors an attribute of the original parameters). This
+    make it easier to handle things (e.g. freeing parameters) on XLA.
+    """
+    if len(params_to_shard) > 0:
+      # When freeing the full parameters, we point their `.data` to this placeholder
+      # (so that the XLA compiler can reuse the memory storage).
+      self._dummy_data_placeholder = torch.zeros(
+          1, device=params_to_shard[0].device)
+
+    # get the module names of each full parameter to shard
+    params_to_shard_set = set(params_to_shard)
+    assert len(params_to_shard_set) == len(params_to_shard), \
+        "params_to_shard should not have dups"
+    full_param_infos = []
+    shared_full_param_memo = {}
+    shared_full_param_infos = []
+    full_params = []
+    for module_name, m in self.named_modules():
+      for n, p in m.named_parameters(recurse=False):
```

Comment:
Will these two lines yield `self._fsdp_wrapped_module.flat_params`? I failed to see how it could by looking at https://github.com/pytorch/pytorch/blob/v1.11.0/torch/nn/modules/module.py#L1668.

Reply:
Yes, these two lines will yield `self._fsdp_wrapped_module.flat_params`. In `XlaFlattenParamsWrapper` the original parameters of the module have been replaced by the flattened parameters.

Reply:
Sure, but how is the original `module` being accessed through `self.named_modules()`?

Reply:
Sure, but how is `self._fsdp_wrapped_module` accessed through `self.named_modules()`?

Reply:
OK I think it's done through https://github.com/pytorch/pytorch/blob/v1.11.0/torch/nn/modules/module.py#L1217. NVM then. Thanks!